services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java

```diff
@@ -373,7 +385,19 @@ private void patchGoalstateForNeighbor(UnicastGoalState unicastGoalState) throws
                 //unicastGoalState.setHostIp(neighborInfo.getHostIp());
                 unicastGoalState.setHostIp(hostIp);
                 unicastGoalState.getGoalStateBuilder().addNeighborStates(neighborState);
-                patchGoalstateForNeighbor(unicastGoalState);
+
+                // use unicastGoalStateTemp object to get patchGoalStates for neighborState update
+                // unicasGoalStateTemp will include subnet_states and a consolidated router_state based on the current neighborState
+                UnicastGoalState unicastGoalStateTemp = new UnicastGoalState();
+                unicastGoalStateTemp.getGoalStateBuilder().addNeighborStates(neighborState);
+                patchGoalstateForNeighbor(networkConfig, unicastGoalStateTemp);
+
+                unicastGoalState.getGoalStateBuilder().addAllSubnetStates(unicastGoalStateTemp.getGoalStateBuilder().getSubnetStatesList());
+                unicastGoalState.getGoalStateBuilder().addAllRouterStates(unicastGoalStateTemp.getGoalStateBuilder().getRouterStatesList());
+                unicastGoalState.getGoalStateBuilder().addAllSubnetStates(multicastGoalState.getGoalStateBuilder().getSubnetStatesList());
+                rebuildRouterState(unicastGoalState.getGoalStateBuilder(), multicastGoalState.getGoalStateBuilder());
+                multicastGoalState.getGoalStateBuilder().addAllSubnetStates(unicastGoalStateTemp.getGoalStateBuilder().getSubnetStatesList());
+                rebuildRouterState(multicastGoalState.getGoalStateBuilder(), unicastGoalStateTemp.getGoalStateBuilder());
```

Comment:
Let us see if line 391-400 could be patched in a standalone method.

Reply:
These lines of code are very depends on the scenario, it's only for the test scenario #5, to patch neghbor_state and rebuild router_state. Unless we have new scenario to use the same sequence of patching (I cannot think of any now), otherwise it's not worth to create a method especially for this case.

Reply:
The unicastGoalStateTemp is a local variable, that's is a variable to get the output of the `patchGoalstateForNeighbor`. Since I made this function for both scenario #4 and #5, so here the unicastGoalStateTemp is necessary to get patched states, otherwise if we use original unicastGoalState or multicaseGoalState, it will mess up both states.

Reply:
Makes sense. Thanks.